storage/tokudb/ha_tokudb.cc

```diff
@@ -3876,31 +3884,33 @@ void ha_tokudb::set_main_dict_put_flags(THD* thd, bool opt_eligible, uint32_t* p
     // to do. We cannot do this if otherwise, because then we lose
     // consistency between indexes
     //
-    if (hidden_primary_key) 
-    {
+    if (hidden_primary_key) {
         *put_flags = old_prelock_flags;
-    }
-    else if (!do_unique_checks(thd, in_rpl_write_rows | in_rpl_update_rows) && !is_replace_into(thd) && !is_insert_ignore(thd))
-    {
+    } else if (!do_unique_checks(thd, in_rpl_write_rows | in_rpl_update_rows) &&
+               !is_replace_into(thd) &&
+               !is_insert_ignore(thd)) {
         *put_flags = old_prelock_flags;
-    }
-    else if (using_ignore_flag_opt && is_replace_into(thd) 
-            && !in_hot_index)
-    {
+    } else if (using_ignore_flag_opt && is_replace_into(thd) && !in_hot_index) {
         *put_flags = old_prelock_flags;
-    }
-    else if (opt_eligible && using_ignore_flag_opt && is_insert_ignore(thd) 
-            && !in_hot_index)
-    {
-        *put_flags = DB_NOOVERWRITE_NO_ERROR | old_prelock_flags;
-    }
-    else 
-    {
+    } else if (opt_eligible && using_ignore_flag_opt &&
+               is_insert_ignore(thd) &&
+               !in_hot_index) {
+        // GL on DB-937 : This is incorrect. The server expects an SE to
+        // return ER_DUP_ENTRY on a dup key hit even when IGNORE is in use.
+        // This is so the server can set the correct warnings on the statement.
+        //*put_flags = DB_NOOVERWRITE_NO_ERROR | old_prelock_flags;
+        *put_flags = DB_NOOVERWRITE | old_prelock_flags;
```

Comment:
The else if condition can be removed and merged with the else now, removing opt_eligible arg from the method too?